### ISSUE TITLE
Tiled gallery block: allow adding images from media gallery after adding images

### ISSUE THREAD

**Issue by simison** [Automattic/wp-calypso#28905]
(Pending merging PR https://github.com/Automattic/wp-calypso/pull/27458)

When adding Tiled gallery block, user can choose to upload or add image via media gallery:

<img width="602" alt="image" src="https://user-images.githubusercontent.com/87168/49143915-c4161380-f304-11e8-8b2b-029a56f212d9.png">

After adding images, user can only choose to upload images:

<img width="622" alt="image" src="https://user-images.githubusercontent.com/87168/49143942-ce381200-f304-11e8-92b3-a4fcd5b3d233.png">

We should follow closely what core gallery does for this section and possibly aim to componentize it: https://github.com/WordPress/gutenberg/pull/9682

Relates to permissions checking: https://github.com/WordPress/gutenberg/blob/bcf55ff095fb9cbfee37aef24c422fbbf9661562/docs/data/data-core.md#hasuploadpermissions

**Comment by sirreal** [comment]
I was implementing this in #30863 where I attempted to unify the empty gallery placeholder and the add media placeholder, but have hit significant roadblocks.

**Comment by simison** [comment]
I think we should piggyback on this implementation once it merges https://github.com/WordPress/gutenberg/pull/12367

**Comment by sirreal** [comment]
> I think we should piggyback on this implementation once it merges [WordPress/gutenberg#12367](https://github.com/WordPress/gutenberg/pull/12367)

Yep, that implementation is essentially what I propose in #30863. A large part of `MediaPlaceholder` is duplicated. I suspect https://github.com/WordPress/gutenberg/pull/9682 to experience the same issues I discovered in #30863.

**Comment by simison** [comment]
Do you think it's related to media gallery bugs fixed in https://github.com/WordPress/gutenberg/pull/12435 ?

**Comment by sirreal** [comment]
No, there are just limitations to the MediaPlaceholder. See the full description: https://github.com/Automattic/wp-calypso/pull/30863#issue-253992640

Specifically, there's no clear way around this limitation:

> * When uploading media via the placeholder button to an existing gallery, it completely replaces the existing images.
> * When uploading media via drag and drop to an existing gallery, it completely replaces the existing images.

Otherwise, the `MediaPlaceholder` implementation is quite nice.

**Comment by simison** [comment]
cc @jorgefilipecosta

**Comment by jorgefilipecosta** [comment]
Hi @simison, @sirreal,
My plan was to after merging https://github.com/WordPress/gutenberg/pull/12367 extract a component that allows rendering the gallery appender. After reading this PR and the related ones https://github.com/Automattic/wp-calypso/pull/30863 I noticed that the gallery appender being implemented in https://github.com/WordPress/gutenberg/pull/12367  is just equivalent to `MediaPlaceholder` without label and icon, with the media library button opening the media library in a different state, and with some custom styles.
It may be possible refactor https://github.com/WordPress/gutenberg/pull/12367  to use  `MediaPlaceholder` in the gallery appender by adding two additional optional props one that enables a custom style and one that allows opening the media library in a specific state (the add to gallery mode).
Feel free to share your thoughts if you think this would be a positive change or if you prefer a different path.

**Comment by sirreal** [comment]
Thanks @jorgefilipecosta! What you outline is aligned with my ideas 👍 

Two things I'd add:

`MediaPlaceholder` needs to take into account that uploaded images, via dialog or drag and drop, should be appended to the current gallery rather than replace it as they do now.

`MediaPlaceholder` includes `DropZone`. Core Gallery and Tiled Gallery remove the appender section when the block is not selected. If we rely on `MediaPlaceholder` and its `DropZone`, we cannot remove it when the block is not selected or we lose the `DropZone`. It would be nice to be able to remove the appender UI but keep the `DropZone` functionality when the block is not selected. In #30863 I've done this with CSS.
